doc/threadsafe_function_ex.md

```diff
@@ -0,0 +1,317 @@
+# ThreadSafeFunctionEx
+
+The `Napi::ThreadSafeFunctionEx` type provides APIs for threads to communicate
+with the addon's main thread to invoke JavaScript functions on their behalf. The
+type is a three-argument templated class, each argument representing the type
+of:
+- `ContextType = std::nullptr_t`: The thread-safe function's context. By default,
+  a TSFN has no context.
+- `DataType = void*`: The data to use in the native callback. By default, a TSFN
+  can accept *any* data type.
+- `Callback = void(*)(Napi::Env, Napi::Function jsCallback, ContextType*,
+  DataType*)`: The callback to run for each item added to the queue. If no
+  `Callback` is given, the API will call the function `jsCallback` with no
+  arguments.
+
+Documentation can be found for an [overview of the API](threadsafe.md), as well
+as [differences between the two thread-safe function
+APIs](threadsafe.md#implementation-differences).
+
+## Methods
+
+### Constructor
+
+Creates a new empty instance of `Napi::ThreadSafeFunctionEx`.
+
+```cpp
+Napi::Function::ThreadSafeFunctionEx<ContextType, DataType, Callback>::ThreadSafeFunctionEx();
+```
+
+### Constructor
+
+Creates a new instance of the `Napi::ThreadSafeFunctionEx` object.
+
+```cpp
+Napi::ThreadSafeFunctionEx<ContextType, DataType, Callback>::ThreadSafeFunctionEx(napi_threadsafe_function tsfn);
```

Comment:
Does this constructor also need `Napi::Env env`? What if the user calls `.Ref()` or `.Unref()`?

Reply:
I believe this is resolved as we discussed last week: TSFN[Ex] have their own implementation of `.Ref()` and `.Unref()`